.github/workflows/build.yml

```
@@ -148,135 +165,233 @@ jobs:
+      id-token: write # for AWS KMS signing (see usage below)
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          sparse-checkout-cone-mode: 'false'
+          sparse-checkout: |
+            .scripts/oci.jq
+            .scripts/provenance.jq
+      - name: Download a single artifact
+        uses: actions/download-artifact@v4
+        with:
+          name: build-oci
+      - name: Verify artifact
+        run: |
+          echo "${{ needs.build.outputs.sha256 }}" sha256sum -c
+          mkdir -p temp
+          tar -xvf temp.tar -C temp
       - name: Configure AWS (for signing)
-        if: fromJSON(steps.json.outputs.json).shouldSign
         # https://github.com/aws-actions/configure-aws-credentials/releases
         uses: aws-actions/configure-aws-credentials@010d0da01d0b5a38af31e9c3470dbfdabdecca3a # v4.0.1
         with:
-          aws-region:     ${{ github.ref_name == 'main' && secrets.AWS_KMS_PROD_REGION   || secrets.AWS_KMS_STAGE_REGION }}
-          role-to-assume: ${{ github.ref_name == 'main' && secrets.AWS_KMS_PROD_ROLE_ARN || secrets.AWS_KMS_STAGE_ROLE_ARN }}
-          # TODO figure out if there's some way we could make our secrets ternaries here more DRY without major headaches 🙈
+          aws-region:     ${{ secrets.AWS_KMS_REGION }}
+          role-to-assume: ${{ secrets.AWS_KMS_ROLE_ARN }}
+      - name: Generate Provenance
+        env:
+          buildJson:      ${{ needs.build.outputs.buildJson }}
+          GITHUB_CONTEXT: ${{ toJson(github) }}
+        run: |
+          image-digest() {
+            local dir="$1/blobs"
+            img=$(
+              grep -R --include "*" '"mediaType":\s"application/vnd.oci.image.layer.' "$dir" \
+                | head -n 1 \
+                | cut -d ':' -f1
+            )
+            [ "$(cat $img | jq -r '.mediaType')" = "application/vnd.oci.image.manifest.v1+json" ] || exit 1
+            echo $img | rev | cut -d '/' -f2,1 --output-delimiter ':' | rev
+          }
+
+          digest=$(image-digest temp)
+
+          echo $buildJson | jq -L.scripts --argjson github '${{ env.GITHUB_CONTEXT }}' --argjson runner '${{ toJson(runner) }}' --arg digest ${digest} '
+            include "provenance";
+            github_actions_provenance($github; $runner; $digest)
+          ' >> provenance.json
       - name: Sign
-        if: fromJSON(steps.json.outputs.json).shouldSign
         env:
-          AWS_KMS_REGION:  ${{ github.ref_name == 'main' && secrets.AWS_KMS_PROD_REGION  || secrets.AWS_KMS_STAGE_REGION }}
-          AWS_KMS_KEY_ARN: ${{ github.ref_name == 'main' && secrets.AWS_KMS_PROD_KEY_ARN || secrets.AWS_KMS_STAGE_KEY_ARN }}
+          AWS_KMS_REGION:  ${{ secrets.AWS_KMS_REGION }}
+          AWS_KMS_KEY_ARN: ${{ secrets.AWS_KMS_KEY_ARN }}
+
+          DOCKER_HUB_USERNAME: ${{ secrets.DOCKER_HUB_USERNAME }}
+          DOCKER_HUB_PASSWORD: ${{ secrets.DOCKER_HUB_PASSWORD }}
```

I created a separate set of secrets for this, since it's a different repository and can have a different security model (so that the signing step doesn't need permission to push to the staging repositories at all) -- it's wired up to the same credentials for now, but I plan to adjust that in the future:

```suggestion
          DOCKER_HUB_USERNAME: ${{ secrets.DOCKER_HUB_SIGNING_USERNAME }}
          DOCKER_HUB_PASSWORD: ${{ secrets.DOCKER_HUB_SIGNING_PASSWORD }}
```